transit.md

```diff
@@ -167,23 +170,164 @@ The Transit client can attempt connections to multiple relays, and uses the
 first one that passes negotiation. Each side combines a locally-configured
 hostname/port (usually "baked in" to the application, and hosted by the
 application author) with additional hostname/port pairs that come from the
-peer. This way either side can suggest the relays to use. The `wormhole`
-application accepts a `--transit-helper tcp:myrelay.example.org:12345`
-command-line option to supply an additional relay. The connection hints
+peer. This way either side can suggest the relays to use. The connection hints
 provided by the Transit instance include the locally-configured relay, along
 with the dynamically-determined direct hints. Both should be delivered to the
 peer.
 
+
+## Canonical abilities encodings
+
+The transit protocol relies on an existing secured (but possibly low-bandwidth)
+communication channel to exchange the abilities and hints. Thus, it has no
+influence over how they are encoded. However, we make a suggestion using JSON
+messages for other protocols to use.
+
+Abilities are encoded as a list, each item having a `type` tag. An ability may
+not appear more than once in the list. Setting `relay-v2` requires also setting
+`relay-v2` for backwards compatibility.
+
+```json
+[
+  {
+    "type": "<string, one of {direct-tcp-v1, relay-v1, tor-tcp-v1}>"
+  }
+]
+```
+
+Example for the full abilities set:
+
+```json
+[
+  { "type": "direct-tcp-v1" },
+  { "type": "relay-v1" },
+  { "type": "relay-v2" },
+  { "type": "tor-tcp-v1" }
+]
+```
+
+## Canonical hint encodings
+
+Hints are encoded as list of objects. Every object contains a `type` field,
+which further determines its encoding.
+
+### `direct-tcp-v1`, `tor-tcp-v1`
+
+```json
+{
+  "type": "direct-tcp-v1" or "tor-tcp-v1",
+  "hostname": <string>,
+  "port": <u16>,
+  "priority": <float; optional>
+}
+```
+
+`hostname` must be compliant with the `host` part of an URL, i.e. it may be an
+IP address or a domain. Furthermore, IPv6 link-local addresses are not supported.
```

Comment:
Definitely like the hostname restriction .. why the caveat about link-local addresses (and why is it v6-only?) Could link-local address be useful during testing? (If they're not, we should exclude IPv4 ones too though, no?)

Reply:
Link-local addresses are bound to the used interface and are thus not compatible across machines even within the same local network. Furthermore, I've made the experience that the required interface suffix is not handled well across the stack (parsers, libraries) — not even Chrom*/Firefox support these URLs; both closed the issue as *wontfix* IIRC.

"Normal" local IPv6 addresses are still allowed of course, although the concept is rather weird since most IPv6 addresses are in a global namespace regardless (no NATting).

Reply:
Okay, maybe link-local addresses are useless .. but aren't v4 ones useless too then?

Reply:
No, since they are scoped per network and not per interface. As soon as two peers are in the same network, it works fine.

Reply:
Okay, I see .. what about 3 Docker containers on the same host, those are still "different interfaces" for IPv6 because each container has its own (virtual) one?

Reply:
I'm not sure, it probably depends on the setup. In any case, we *always* have "normal" IPv6 addresses: we need internet access (or at least a network) to access the rendezvous server.